lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -316,8 +317,18 @@ def _find_matches(self, requirements):
         # The fqcn is guaranteed to be the same
         version_req = "A SemVer-compliant version or '*' is required. See https://semver.org to learn how to compose it correctly. "
         version_req += "This is an issue with the collection."
+
+        # If we're upgrading collections, we can't calculate preinstalled_candidates until the latest matches are found.
+        # Otherwise, we can potentially avoid a Galaxy API call by doing this first.
+        preinstalled_candidates = None
```

Comment:
I think this should remain of the same type throughout the whole run time.
```suggestion
        preinstalled_candidates = set()
```